File: seed/rust-sdk/trace/src/types/variable_value.rs
use crate::map_value::MapValue;
use crate::binary_tree_value::BinaryTreeValue;
use crate::singly_linked_list_value::SinglyLinkedListValue;
use crate::doubly_linked_list_value::DoublyLinkedListValue;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type")]
pub enum VariableValue {
        IntegerValue {
            value: i32,
        },

        BooleanValue {
            value: bool,
        },

        DoubleValue {
            value: f64,
        },

        StringValue {
            value: String,
        },

        CharValue {
            value: String,
        },

        MapValue {
            #[serde(flatten)]
            data: MapValue,
        },

        ListValue {
            value: Vec<VariableValue>,
        },

        BinaryTreeValue {
            #[serde(flatten)]
            data: BinaryTreeValue,
        },

        SinglyLinkedListValue {
            #[serde(flatten)]
            data: SinglyLinkedListValue,
        },

        DoublyLinkedListValue {
            #[serde(flatten)]
            data: DoublyLinkedListValue,
        },

        NullValue,
}

File: seed/ruby-sdk-v2/trace/lib/seed/commons/variable_value.rb
# frozen_string_literal: true

module Seed
    module Types
        class VariableValue < Internal::Types::Union

            discriminant :type

            member -> { Integer }, key: "INTEGER_VALUE"
            member -> { Internal::Types::Boolean }, key: "BOOLEAN_VALUE"
            member -> { Integer }, key: "DOUBLE_VALUE"
            member -> { String }, key: "STRING_VALUE"
            member -> { String }, key: "CHAR_VALUE"
            member -> { Seed::Commons::MapValue }, key: "MAP_VALUE"
            member -> { Internal::Types::Array[Seed::Commons::VariableValue] }, key: "LIST_VALUE"
            member -> { Seed::Commons::BinaryTreeValue }, key: "BINARY_TREE_VALUE"
            member -> { Seed::Commons::SinglyLinkedListValue }, key: "SINGLY_LINKED_LIST_VALUE"
            member -> { Seed::Commons::DoublyLinkedListValue }, key: "DOUBLY_LINKED_LIST_VALUE"
            member -> { Object }, key: "NULL_VALUE"
    end
end
